File: src/net/merayen/merasynth/ui/objects/components/parameterslider/Button.java
package net.merayen.merasynth.ui.objects.components.parameterslider;

import net.merayen.merasynth.ui.Point;
import net.merayen.merasynth.ui.event.IEvent;
import net.merayen.merasynth.ui.objects.UIObject;
import net.merayen.merasynth.ui.util.MouseHandler;

public class Button extends UIObject {
	
	public interface IHandler {
		public void onClick();
	}

	public String label;

	public float width = 5f;
	public float height = 1.5f;

	private IHandler handler; 
	private MouseHandler mousehandler;
	private boolean mouse_down;
	private boolean mouse_over;

	protected void onInit() {
		mousehandler = new MouseHandler(this);
		mousehandler.setHandler(new MouseHandler.IMouseHandler() {
			
			@Override
			public void onMouseUp(Point position) {
				if(mouse_down && handler != null)
					handler.onClick();

				mouse_down = false;
			}
			
			@Override
			public void onMouseOver() {
				mouse_over = true;
			}
			
			@Override
			public void onMouseOut() {
				mouse_over = false;
			}
			
			@Override
			public void onMouseMove(Point position) {}
			
			@Override
			public void onMouseDrop(Point start_point, Point offset) {}
			
			@Override
			public void onMouseDrag(Point start_point, Point offset) {}
			
			@Override
			public void onMouseDown(Point position) {
				mouse_down = true;
			}
			
			@Override
			public void onGlobalMouseMove(Point global_position) {}
			
			@Override
			public void onGlobalMouseUp(Point global_position) {
				mouse_down = false;
			}
		});
	}

	protected void onDraw() {
		draw.setColor(50, 50, 50);
		draw.fillRect(0, 0, width, height);

		if(mouse_down && mouse_over)
			draw.setColor(80, 80, 80);
		else
			draw.setColor(120, 120, 120);
		draw.fillRect(0.1f, 0.1f, width - 0.2f, height - 0.2f);

		draw.setColor(200, 200, 200);
		float text_width = draw.getTextWidth(label);
		draw.text(label, (float)(width/2 - text_width/2), 1f);

		super.onDraw();
	}

	protected void onEvent(IEvent event) {
		mousehandler.handle(event);
	}

	public void setHandler(IHandler handler) {
		this.handler = handler;
	}
}

File: src/main/kotlin/net/merayen/elastic/ui/objects/components/buttons/Button.kt
package net.merayen.elastic.ui.objects.components.buttons

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.FlexibleDimension
import net.merayen.elastic.ui.MutableColor
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.util.MouseHandler
import net.merayen.elastic.util.MutablePoint

open class Button(var label: String = "") : UIObject(), FlexibleDimension {
	override var layoutWidth = 50f
	override var layoutHeight = 15f
	var autoDimension = true
	var fontSize = 12f
	var textColor = MutableColor()
	var backgroundColor = MutableColor(0.5f, 0.5f, 0.5f)

	var disabled = false

	var handler: IHandler? = null
	private var mousehandler: MouseHandler? = null
	private var mouseDown: Boolean = false
	private var mouseOver: Boolean = false

	interface IHandler {
		fun onClick()
	}

	override fun onInit() {
		mousehandler = MouseHandler(this)
		mousehandler!!.setHandler(object : MouseHandler.Handler() {

			override fun onMouseUp(position: MutablePoint) {
				if (mouseDown && handler != null)
					handler!!.onClick()

				mouseDown = false
			}

			override fun onMouseOver() {
				mouseOver = true
			}

			override fun onMouseOut() {
				mouseOver = false
			}

			override fun onMouseDown(position: MutablePoint) {
				mouseDown = true
			}

			override fun onGlobalMouseUp(global_position: MutablePoint) {
				mouseDown = false
			}
		})
	}

	override fun onDraw(draw: Draw) {
		draw.setFont("Geneva", fontSize)

		val textWidth = draw.getTextWidth(label)
		if (autoDimension) {
			layoutWidth = textWidth + 10
			layoutHeight = fontSize * 1.5f
		}

		draw.setColor(0.2f, 0.2f, 0.2f, if (disabled) 0.5f else 1f)
		draw.fillRect(0f, 0f, layoutWidth, layoutHeight)

		if (mouseDown && mouseOver)
			draw.setColor(
				backgroundColor.red + (1f - backgroundColor.red) * 0.5f,
				backgroundColor.green + (1f - backgroundColor.green) * 0.5f,
				backgroundColor.blue + (1f - backgroundColor.blue) * 0.5f,
				if (disabled) 0.5f else 1f
			)
		else
			draw.setColor(backgroundColor.red, backgroundColor.green, backgroundColor.blue, if (disabled) 0.5f else 1f)

		draw.fillRect(1f, 1f, layoutWidth - 2f, layoutHeight - 2f)

		draw.setColor(textColor.red, textColor.green, textColor.blue, if (disabled) 0.5f else 1f)
		if (mouseDown && mouseOver)
			draw.text(label, layoutWidth / 2 - textWidth / 2, fontSize + fontSize / 5)
		else
			draw.text(label, layoutWidth / 2 - textWidth / 2, fontSize)
	}

	override fun onEvent(event: UIEvent) {
		if (!disabled) {
			mousehandler!!.handle(event)
		} else {
			mouseDown = false
			mouseOver = false
		}
	}
}